scripts/bin/update-packages

```diff
@@ -148,22 +151,130 @@ _fetch_and_cache_tags() {
+	local __GITHUB_PACKAGES=()
 	for repo_dir in $(jq --raw-output 'del(.pkg_format) | keys | .[]' "${TERMUX_SCRIPTDIR}/repo.json"); do
 		for pkg_dir in "${repo_dir}"/*; do
-			! quiet _should_update "${pkg_dir}" && continue # Skip if not needed.
+			quiet _should_update "${pkg_dir}" || continue # Skip if not needed.
 			grep -q '^termux_pkg_auto_update' "${pkg_dir}/build.sh" && continue # Skip if package has custom auto-update
-			__PACKAGES+=("${pkg_dir}")
+			if grep '^TERMUX_PKG_SRCURL=' "${pkg_dir}/build.sh" | grep -q 'github.com'; then
+				__GITHUB_PACKAGES+=("${pkg_dir}")
+			else
+				__PACKAGES+=("${pkg_dir}")
+			fi
 		done
 	done
 
+	local GITHUB_GRAPHQL_QUERIES=() COUNTER=0
+	declare -A __GITHUB_GRAPHQL_PACKAGES=()
+	# Ignore non-constant sources
+	# shellcheck disable=SC1091
+	for pkg_dir in "${__GITHUB_PACKAGES[@]}"; do
+		local PKG_SRCURL TAG_TYPE OWNER REPO
+		read -r PKG_SRCURL TAG_TYPE < <(
+			set +u
+			source "${pkg_dir}/build.sh"
+			echo "${TERMUX_PKG_SRCURL} ${TERMUX_PKG_UPDATE_TAG_TYPE}"
+		)
+
+		read -r OWNER REPO < <(sed -E 's|^(git\+)?https?://github.com/([^/]+)/([^/]+).*|\2 \3|' <<< "$PKG_SRCURL")
+
+		if [[ -z "${TAG_TYPE}" ]]; then # If not set, then decide on the basis of url.
+			if [[ "${PKG_SRCURL:0:4}" == "git+" ]]; then
+				# Get newest tag.
+				TAG_TYPE="newest-tag"
+			else
+				# Get the latest release tag.
+				TAG_TYPE="latest-release-tag"
+			fi
+		fi
+
+		# This holds the source repositories, and query type for each repo we need to query.
+		# Newlines need to be escaped for the query parser
+		GITHUB_GRAPHQL_QUERIES+=( "_$((COUNTER++)): repository(owner: \\\"${OWNER}\\\", name: \\\"${REPO}\\\") { ..._${TAG_TYPE//-/_} }" )
+
+		unset PKG_SRCURL TAG_TYPE OWNER REPO
+	done
+
+	# Batch size for fetching tags, 100 seems to work consistently.
+	local BATCH BATCH_SIZE=100
+	echo "// Begin test: $(date)" > /tmp/query-12345
+	for (( BATCH = 0; ${#GITHUB_GRAPHQL_QUERIES[@]} >= BATCH_SIZE * BATCH ; BATCH++ )); do
+
+		echo "Starting batch $BATCH at: ${GITHUB_GRAPHQL_QUERIES[$BATCH * $BATCH_SIZE]//\\/}"
+
+		# JSON strings cannot contain tabs or newlines
+		# so shutup shellcheck complaining about escapes in single quotes
+		local QUERY
+
+		# Start the GraphQL query with our two fragments for getting the latest tag from a release, and from refs/tags
+		# These are defined only if needed, so this one is for '_latest_release_tag'
+		grep -q '_latest_release_tag' <<< "${GITHUB_GRAPHQL_QUERIES[@]:$BATCH * $BATCH_SIZE:$BATCH_SIZE}" && {
+			QUERY+="$(printf '%s\n' \
+			'fragment _latest_release_tag on Repository {' \
+			'  latestRelease { tagName }' \
+			'}')"
+		}
+
+		grep -q '_newest_tag' <<< "${GITHUB_GRAPHQL_QUERIES[@]:$BATCH * $BATCH_SIZE:$BATCH_SIZE}" && {
+			QUERY+="$(printf '%s\n' \
+			'fragment _newest_tag on Repository {' \
+			'  refs( refPrefix: \"refs/tags/\" first: 1 orderBy: {field: TAG_COMMIT_DATE, direction: DESC}) {' \
+			'    nodes { name }' \
+			'  }' \
+			'}')"
+		}
```

Comment:
We can use heredoc or multiline string syntax and drop newlines, hard tabs and unnecessary spaces during HTTP request composition (`tr -d '\n\t' | sed 's/[ ][ ]*/ /g'`?).

Reply:
I'd prefer to keep this simple.
And introducing heredoc semantics, `tr` and `sed` into this decidedly does not do that.
It's unfortunately a bit messy either way.

Reply:
Actually we already use this solution in auto-update code. [Here](https://github.com/termux/termux-packages/blob/7daa0efe4a7a0cef755a225a2c428ec7e30f01b1/scripts/updates/api/termux_github_api_get_tag.sh#L54-L74).